pkg/cloud/services/networking/port.go

```diff
@@ -0,0 +1,251 @@
+		AdminStateUp:        portOpts.AdminStateUp,
+		MACAddress:          portOpts.MACAddress,
+		TenantID:            portOpts.TenantID,
+		ProjectID:           portOpts.ProjectID,
+		SecurityGroups:      securityGroups,
+		AllowedAddressPairs: addressPairs,
+		FixedIPs:            fixedIPs,
+	}
+
+	if portOpts.DisablePortSecurity != nil {
+		portSecurity := !*portOpts.DisablePortSecurity
+		createOpts = portsecurity.PortCreateOptsExt{
+			CreateOptsBuilder:   createOpts,
+			PortSecurityEnabled: &portSecurity,
+		}
+	}
+
+	createOpts = portsbinding.CreateOptsExt{
+		CreateOptsBuilder: createOpts,
+		HostID:            portOpts.HostID,
+		VNICType:          portOpts.VNICType,
+		Profile:           getPortProfile(portOpts.Profile),
+	}
+
+	port, err := s.client.CreatePort(createOpts)
+	if err != nil {
+		record.Warnf(eventObject, "FailedCreatePort", "Failed to create port %s: %v", portName, err)
+		return nil, err
+	}
+
+	record.Eventf(eventObject, "SuccessfulCreatePort", "Created port %s with id %s", port.Name, port.ID)
+	if portOpts.Trunk != nil && *portOpts.Trunk {
+		trunk, err := s.getOrCreateTrunk(eventObject, clusterName, port.Name, port.ID)
+		if err != nil {
+			record.Warnf(eventObject, "FailedCreateTrunk", "Failed to create trunk for port %s: %v", portName, err)
+			return nil, err
+		}
+		if err = s.replaceAllAttributesTags(eventObject, trunk.ID, tags); err != nil {
+			record.Warnf(eventObject, "FailedReplaceTags", "Failed to replace trunk tags %s: %v", portName, err)
+			return nil, err
+		}
+	}
+
+	return port, nil
+}
+
+func getPortProfile(p map[string]string) map[string]interface{} {
+	portProfile := make(map[string]interface{})
+	for k, v := range p {
+		portProfile[k] = v
+	}
+	// We need return nil if there is no profiles
+	// to have backward compatible defaults.
+	// To set profiles, your tenant needs this permission:
+	// rule:create_port and rule:create_port:binding:profile
+	if len(portProfile) == 0 {
+		return nil
+	}
+	return portProfile
+}
+
+func (s *Service) DeletePort(eventObject runtime.Object, portID string) error {
+	port, err := s.getPort(portID)
+	if err != nil {
+		return err
+	}
+	if port == nil {
+		return nil
+	}
+
+	err = util.PollImmediate(retryIntervalPortDelete, timeoutPortDelete, func() (bool, error) {
+		err := s.client.DeletePort(port.ID)
+		if err != nil {
+			if capoerrors.IsNotFound(err) {
+				record.Eventf(eventObject, "SuccessfulDeletePort", "Port %s with id %d did not exist", port.Name, port.ID)
+			}
+			if capoerrors.IsRetryable(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		return true, nil
+	})
+	if err != nil {
+		record.Warnf(eventObject, "FailedDeletePort", "Failed to delete port %s with id %s: %v", port.Name, port.ID, err)
+		return err
+	}
+
+	record.Eventf(eventObject, "SuccessfulDeletePort", "Deleted port %s with id %s", port.Name, port.ID)
+	return nil
+}
+
+func (s *Service) DeletePorts(eventObject runtime.Object, nets []servers.Network) error {
```

Comment:
nit: I personally wouldn't have moved this method. Its function is to iterate over servers.Network, which is a compute thing. Honestly I'd inline it in createInstance.